src/libraries/System.Text.RegularExpressions/gen/UpgradeToRegexGeneratorAnalyzer.cs

```diff
@@ -53,7 +53,7 @@ public override void Initialize(AnalysisContext context)
                 // Validate that the project is not using top-level statements, since if it were, the code-fixer
                 // can't easily convert to the source generator without having to make the program not use top-level
                 // statements any longer.
```

Comment:
I don't understand why it's not easy? I thought this should be doable with keeping top-level statement. But I haven't spent time to try to do it myself.

Reply:
The main issue here is that in order to use the source generator today, you need to declare the containing type as `partial` since the `RegexGenerator` attributed method will be implemented in a separate file by the source generator. The issue with Top-level-statements is that we don't control the declaration of `Program` as that is done by the compiler, so I believe the only way would be to transform the program to not use top-level-statements any more. This felt like more work than we would want the code-fixer to do since it may not always be what the user wants, so that is why for now we are just not emitting a diagnostic in this case.

Reply:
@joperezr You can declare a partial `Program` with top-level statements. This wasn't in the early design of the feature, but has changed at some point.

The following code should be valid, I think.

```csharp
using System;

M();

partial class Program
{
    public void M() => Console.WriteLine();
}
```

Even if the fixer can't do something, it's not uncommon that a diagnostic doesn't have an associated codefix if the message is clear enough for the user to fix it himself.

Reply:
> Even if the fixer can't do something, it's not uncommon that a diagnostic doesn't have an associated codefix if the message is clear enough for the user to fix it himself.

I certainly agree with this. Even if we avoid reporting diagnostics in top-level statements, I don't see why we should avoid documents elsewhere in the solution.